plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p3order/LayerSweepCrossingMinimizer.java

```diff
@@ -226,19 +230,83 @@ private int minimizeCrossingsWithCounter(final GraphInfoHolder gData) {
+        int wrongModelOrder = 0;
+        for (LNode[] layer : layers) {
+            ModelOrderNodeComparator comp = new ModelOrderNodeComparator(
+                    previousLayer == -1 ? layers[0] : layers[previousLayer], strategy, DummyNodeStrategy.EQUAL);
+            for (int i = 0; i < layer.length; i++) {
+                for (int j = i + 1; j < layer.length; j++) {
+                    if (comp.compare(layer[i], layer[j]) > 0) {
+                        wrongModelOrder++;
+                    }
+                }
+            }
+            previousLayer++;
+        }
+        return wrongModelOrder;
+    }
+
+    /**
+     * Compares all ports in a each layer and counts how often they are not in model order.
+     * This requires that the {@code SortByInputModelProcessor} ran previously.
+     * @param layers layers to check
+     * @return The number of model order conflicts
+     */
+    private int countModelOrderPortChanges(final LNode[][] layers) {
+        int previousLayer = -1;
+        int wrongModelOrder = 0;
+        for (LNode[] layer : layers) {
+            for (LNode lNode : layer) {
+                Comparator<LPort> comp = new ModelOrderPortComparator(
+                        previousLayer == -1 ? layers[0] : layers[previousLayer],
+                                lNode.getProperty(InternalProperties.TARGET_NODE_MODEL_ORDER));
+                for (int i = 0; i < lNode.getPorts().size(); i++) {
+                    for (int j = i + 1; j < lNode.getPorts().size(); j++) {
+                        if (comp.compare(lNode.getPorts().get(i), lNode.getPorts().get(j)) > 0) {
+                            wrongModelOrder++;
+                        }
+                    }
+                }
+            }
+            previousLayer++;
+        }
+        return wrongModelOrder;
+    }
 
     /*
      * We only need to count crossings below the current graph and also only if they are marked as to be processed
      * hierarchically.
      */
-    private int countCurrentNumberOfCrossings(final GraphInfoHolder currentGraph) {
-        int totalCrossings = 0;
+    private double countCurrentNumberOfCrossings(final GraphInfoHolder currentGraph) {
+        double totalCrossings = 0;
         Deque<GraphInfoHolder> countCrossingsIn = new ArrayDeque<>();
         countCrossingsIn.push(currentGraph);
         while (!countCrossingsIn.isEmpty()) {
             GraphInfoHolder gD = countCrossingsIn.pop();
-            totalCrossings +=
-                    gD.crossCounter().countAllCrossings(gD.currentNodeOrder());
+            double modelOrderInfluence = 0;
+            // Additionally add changes to the model order to the number of crossings.
+            // The influence of port and node order can be configured.
+            if (currentGraph.lGraph()
+                    .getProperty(LayeredOptions.CONSIDER_MODEL_ORDER_STRATEGY) != OrderingStrategy.NONE) {
+                modelOrderInfluence += currentGraph.lGraph()
```

Comment:
Could you retrieve the layout option beforehand? Would make the line more readable.

Reply:
Ok.